PROGRESS.md

```diff
@@ -0,0 +1,406 @@
+```
+task inner {
+  Int i
+  command { echo ${i} }
+  output {
+    Int out = read_int(stdout()) + 1
+  }
+}
+
+task outer {
+  Array[Array[Int]] matrix
+  command { cat ${write_tsv(matrix)} }
+  output {
+    String tsv = read_string(stdout())
+  }
+}
+
+workflow w {
+  Array[Array[Int]] array = [[0,1,2],[3,4,5],[6,7,8]]
+
+  scatter(i in array) {
+    scatter(j in i) {
+      call inner {input: i=j}
+    }
+  }
+
+  # inner.out should be [[1,2,3],[4,5,6],[7,8,9]] (???)
+  call outer {input: matrix=inner.out}
+}
+```
+
+Now doing some graph and lexical operations on the `call inner`:
+
+```
+val inner = namespace.calls.find(_.unqualifiedName == "inner").get
+
+println(inner.ancestry.map(_.fullyQualifiedNameWithIndexScopes))
+// List(w.$scatter_0.$scatter_1, w.$scatter_0, w, )
+
+println(inner.upstream.map(_.fullyQualifiedNameWithIndexScopes))
+// Set(w.$scatter_0.$scatter_1)
+
+println(inner.downstream.map(_.fullyQualifiedNameWithIndexScopes))
+// Set(w.outer)
+
+println(inner.resolveVariable("j").map(_.fullyQualifiedNameWithIndexScopes))
+// Some(w.$scatter_0.$scatter_1)
+
+println(upstreamRecursive(inner).map(_.fullyQualifiedNameWithIndexScopes))
+// Set(w.$scatter_0.$scatter_1, w.$scatter_0, w.array)
+```
+
+## Fully-Qualified Names
+
+adder.wdl
+
+```
+task adder {
+  Int i
+  command { echo ${i} }
+  output { Int o = read_int(stdout()) + 1 }
+}
+
+workflow add_workflow {
+  call adder
+}
+```
+
+main.wdl
+
+```
+import "adder.wdl" as ns
+
+workflow main {
+  call ns.adder
+}
+```
+
+List of FQNs in `main.wdl`
+
+|Fully-Qualified Name|
+|--------------------|
+|`main`|
+|`main.adder`|
+|`main.adder.i`|
+|`ns`|
+|`ns.adder`|
+|`ns.adder.i`|
+|`ns.add_workflow`|
+|`ns.add_workflow.adder`|
+|`ns.add_workflow.adder.i`|
+|``|
+
+> **TODO**: `main.adder.o` should also be a valid FQN
+
+Also detects name collisions:
+
+```
+import "foobar" as ns
+Int main = 2
+workflow main {
+  call ns.adder
+}
+```
+
+Results in:
+
+```
+ERROR: Sibling nodes have conflicting names:
+
+Declaration defined here (line 2, col 1):
+
+Int main = 2
+^
+
+Workflow statement defined here (line 3, col 10):
+
+workflow main {
+         ^
+```
+
+## Global Variables
+
+```
+Int global = 2
+
+task t {
+  Int i = global + 2
+  command { echo ${i} ${global} }
+  runtime {
+    memory: global + "GB"
+  }
+  output { String o = read_string(stdout()) }
+}
+
+workflow w {
+  Int x = global + 2
+  call t
+}
+```
+
+|Fully-Qualified Name|
+|--------------------|
+|`global`|
+|`t`|
+|`t.i`|
+|`w`|
+|`w.x`|
+|`w.t`|
+|`w.t.i`|
+
+## Test Suite Redesign
+
+Tests for lexical relationships, graph relationships, and FQNs have been refactored to be auto-generating.
+
+* `src/test/cases` contains 1 directory per WDL to test.  Each directory **must** contain a `test.wdl` file.
+* `src/test/cases/*/*.json` are the expected results.  For example, `src/test/cases/nested_scatter/downstream.json` contains a JSON object where keys are FQNs in `./test.wdl` and values are array of downstream FQNs.
```

Comment:
This is really custom made for wdl4s scope hierarchy / depedency testing and doesn't involve at all json workflow inputs, but I can rename it for clarity.